torchsnapshot/storage_plugins/fsspec.py

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import asyncio
+import io
+import os
+
+import fsspec
+
+from torchsnapshot.io_types import ReadIO, StoragePlugin, WriteIO
+
+__all__ = ["FSSpecStoragePlugin"]
+
+
+class FSSpecStoragePlugin(StoragePlugin):
+    def __init__(self, root: str) -> None:
+        root_items = root.split("://")
+        if len(root_items) != 2:
+            raise ValueError("only protocol://path is supported by fsspec plugin")
+        protocol, self.root = root_items
+        if not protocol.startswith("fsspec-"):
+            raise ValueError(
+                f"Invalid protocol: {protocol}, Only fsspec-* protocols are supported"
+            )
+        self.fs = fsspec.filesystem(protocol=protocol.removeprefix("fsspec-"))
+        self._session = None
+        self._lock = asyncio.Lock()
+
+    async def _init_session(self) -> None:
+        async with self._lock:
+            if self._session is None:
+                self._session = await self.fs.set_session()
+
+    async def write(self, write_io: WriteIO) -> None:
+        await self._init_session()
+        path = os.path.join(self.root, write_io.path)
+        await self.fs._pipe_file(path, bytes(write_io.buf))
+
+    async def read(self, read_io: ReadIO) -> None:
```

Comment:
If `byte_range` is present in `ReadIO`, we need to honor it.

IIUC, not all `fsspec` plugins support `seek` (or efficient `seek`). We could provide an inefficient backoff here (we can look into that later).
```
@dataclass
class ReadIO:
    path: str
    buf: io.BytesIO = field(default_factory=io.BytesIO)
    byte_range: Optional[Tuple[int, int]] = None
```